utils/sqlcommenter/src/SqlCommenter.hs

```diff
@@ -0,0 +1,268 @@
+lookupSqlCommenterAttributes ctxt = case Ctxt.lookup sqlCommenterKey ctxt of
+  Nothing -> mempty
+  Just attrs -> attrs
+
+getSqlCommenterAttributes :: IO (Map Text Text)
+getSqlCommenterAttributes = lookupSqlCommenterAttributes <$> TL.getContext
+
+addTraceDataToAttributes :: Core.Span -> Map Text Text -> IO (Map Text Text)
+addTraceDataToAttributes span attrs = do
+  (traceparent, tracestate) <- encodeSpanContext span
+  pure $ attrs
+    & M.insert "traceparent" (unsafeConvert traceparent)
+    & M.insert "tracestate" (unsafeConvert tracestate)
+  where
+    unsafeConvert = B.toText . B.unsafeFromByteString
+
+getSqlCommenterAttributesWithTraceData :: IO (Map Text Text)
+getSqlCommenterAttributesWithTraceData = do
+  ctxt <- TL.getContext
+  let attrs = lookupSqlCommenterAttributes ctxt
+  case Ctxt.lookupSpan ctxt of
+    Nothing -> pure attrs
+    Just span -> addTraceDataToAttributes span attrs
+
+-- Represents a set of Word8 values using four Word64 values for 256 bits.
+data Word8Set = Word8Set !Word64 !Word64 !Word64 !Word64 deriving (Eq, Show)
+
+-- | Unsafe conversion between 'Char' and 'Word8'. This is a no-op and
+-- silently truncates to 8 bits Chars > '\255'.
+c2w :: Char -> Word8
+c2w = fromIntegral . ord
+{-# INLINE c2w #-}
+
+-- Creates an empty Word8Set.
+empty :: Word8Set
+empty = Word8Set 0 0 0 0
+
+-- Helper function to determine which Word64 to use and the bit position.
+selectWord64 :: Word8 -> (Word8, Word8)
+selectWord64 val = val `divMod` 64
+
+-- Adds a Word8 value to the set.
+insert :: Word8 -> Word8Set -> Word8Set
+insert val (Word8Set w0 w1 w2 w3) = case selectWord64 val of
+    (0, pos) -> Word8Set (setBit w0 $ fromIntegral pos) w1 w2 w3
+    (1, pos) -> Word8Set w0 (setBit w1 $ fromIntegral pos) w2 w3
+    (2, pos) -> Word8Set w0 w1 (setBit w2 $ fromIntegral pos) w3
+    (3, pos) -> Word8Set w0 w1 w2 (setBit w3 $ fromIntegral pos)
+    _        -> error "Word8 value out of bounds"
+
+-- Removes a Word8 value from the set.
+delete :: Word8 -> Word8Set -> Word8Set
+delete val (Word8Set w0 w1 w2 w3) = case selectWord64 val of
+    (0, pos) -> Word8Set (clearBit w0 $ fromIntegral pos) w1 w2 w3
+    (1, pos) -> Word8Set w0 (clearBit w1 $ fromIntegral pos) w2 w3
+    (2, pos) -> Word8Set w0 w1 (clearBit w2 $ fromIntegral pos) w3
+    (3, pos) -> Word8Set w0 w1 w2 (clearBit w3 $ fromIntegral pos)
+    _        -> error "Word8 value out of bounds"
+
+-- Checks if a Word8 value is in the set.
+member :: Word8 -> Word8Set -> Bool
+member val (Word8Set w0 w1 w2 w3) = case selectWord64 val of
+    (0, pos) -> testBit w0 $ fromIntegral pos
+    (1, pos) -> testBit w1 $ fromIntegral pos
+    (2, pos) -> testBit w2 $ fromIntegral pos
+    (3, pos) -> testBit w3 $ fromIntegral pos
+    _        -> error "Word8 value out of bounds"
+
+-- SQL Commenter spec wants them escaped with a slash, but this should
+-- probably solve the same issue
+unreservedQS :: Word8Set
+unreservedQS = foldr insert SqlCommenter.empty $ map c2w "-_.~'"
```

Comment:
This builds a data structure that can perform fast checks to determine whether to escape a character or not for a URL. This is similar to @ekmett's charset package, but with a much more constrained purposed, and is intended to avoid incurring extra dependencies. https://hackage.haskell.org/package/charset-0.3.11/docs/Data-CharSet.html